File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/shops/PurchaseDialogQuestContent.java
package com.habitrpg.android.habitica.ui.views.shops;

import android.content.Context;
import android.support.annotation.Nullable;
import android.text.Layout;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RatingBar;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.models.inventory.QuestCollect;
import com.habitrpg.android.habitica.models.inventory.QuestContent;
import com.habitrpg.android.habitica.models.inventory.QuestDropItem;
import com.habitrpg.android.habitica.models.shops.ShopItem;
import com.habitrpg.android.habitica.ui.helpers.DataBindingUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * Created by phillip on 24.07.17.
 */

public class PurchaseDialogQuestContent extends PurchaseDialogContent {

    @BindView(R.id.questDetailView)
    View questDetailView;
    @BindView(R.id.questTypeTextView)
    TextView questTypeTextView;
    @BindView(R.id.boss_health_view)
    View bossHealthView;
    @BindView(R.id.boss_health_text)
    TextView bossHealthTextView;
    @BindView(R.id.quest_collect_view)
    View questCollectView;
    @BindView(R.id.quest_collect_text)
    TextView questCollectTextView;
    @BindView(R.id.quest_difficulty_view)
    RatingBar questDifficultyView;
    @BindView(R.id.rage_meter_view)
    View rageMeterView;
    @BindView(R.id.rewardsList)
    ViewGroup rewardsList;
    @BindView(R.id.ownerRewardsTitle)
    View ownerRewardsTitle;
    @BindView(R.id.ownerRewardsList)
    ViewGroup ownerRewardsList;

    public PurchaseDialogQuestContent(Context context) {
        super(context);
    }

    public PurchaseDialogQuestContent(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected int getViewId() {
        return R.layout.dialog_purchase_content_quest;
    }

    @Override
    public void setItem(ShopItem item) {
        super.setItem(item);
    }

    public void setQuestContent(QuestContent questContent) {

        if (questContent.isBossQuest()) {
            questTypeTextView.setText(R.string.boss_quest);
            questCollectView.setVisibility(View.GONE);
            bossHealthTextView.setText(String.valueOf(questContent.getBoss().hp));
            if (questContent.getBoss().hasRage()) {
                rageMeterView.setVisibility(View.VISIBLE);
            }
            questDifficultyView.setRating(questContent.getBoss().str);
        } else{
            questTypeTextView.setText(R.string.collection_quest);
            List<String> collectionList = new ArrayList<>();
            for (QuestCollect collect : questContent.getCollect()) {
                collectionList.add(collect.count + " " + collect.text);
            }
            questCollectTextView.setText(TextUtils.join(", ", collectionList));

            bossHealthView.setVisibility(View.GONE);

            questDifficultyView.setRating(1);
        }

        questDetailView.setVisibility(View.VISIBLE);

        LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if (questContent.getDrop().getItems() != null) {
            for (QuestDropItem item : questContent.getDrop().getItems()) {
                if (!item.isOnlyOwner()) {
                    addRewardsRow(inflater, item, rewardsList);
                }
            }

            boolean hasOwnerRewards = false;
            for (QuestDropItem item : questContent.getDrop().getItems()) {
                if (item.isOnlyOwner()) {
                    addRewardsRow(inflater, item, ownerRewardsList);
                    hasOwnerRewards = true;
                }
            }
            if (!hasOwnerRewards) {
                ownerRewardsTitle.setVisibility(View.GONE);
                ownerRewardsList.setVisibility(View.GONE);
            }

            if (questContent.getDrop().exp > 0) {
                ViewGroup view = (ViewGroup) inflater.inflate(R.layout.row_quest_reward, rewardsList, false);
                SimpleDraweeView imageView = (SimpleDraweeView) view.findViewById(R.id.imageView);
                TextView titleTextView = (TextView) view.findViewById(R.id.titleTextView);
                titleTextView.setText(getContext().getString(R.string.experience_reward, questContent.getDrop().exp));
                rewardsList.addView(view);
            }

            if (questContent.getDrop().gp > 0) {
                ViewGroup view = (ViewGroup) inflater.inflate(R.layout.row_quest_reward, rewardsList, false);
                SimpleDraweeView imageView = (SimpleDraweeView) view.findViewById(R.id.imageView);
                TextView titleTextView = (TextView) view.findViewById(R.id.titleTextView);
                titleTextView.setText(getContext().getString(R.string.gold_reward, questContent.getDrop().gp));
                rewardsList.addView(view);
            }
        }
    }

    private void addRewardsRow(LayoutInflater inflater, QuestDropItem item, ViewGroup containerView) {
        ViewGroup view = (ViewGroup) inflater.inflate(R.layout.row_quest_reward, containerView, false);
        SimpleDraweeView imageView = (SimpleDraweeView) view.findViewById(R.id.imageView);
        TextView titleTextView = (TextView) view.findViewById(R.id.titleTextView);
        DataBindingUtils.loadImage(imageView, item.getImageName());
        if (item.getCount() > 1) {
            titleTextView.setText(getContext().getString(R.string.quest_reward_count, item.getText(), item.getCount()));
        } else {
            titleTextView.setText(item.getText());
        }
        containerView.addView(view);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/shops/PurchaseDialogQuestContent.kt
package com.habitrpg.android.habitica.ui.views.shops

import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.databinding.DialogPurchaseContentQuestBinding
import com.habitrpg.android.habitica.models.inventory.QuestContent
import com.habitrpg.android.habitica.models.inventory.QuestDropItem
import com.habitrpg.android.habitica.ui.views.HabiticaIconsHelper
import com.habitrpg.common.habitica.extensions.layoutInflater
import com.habitrpg.common.habitica.extensions.loadImage
import com.habitrpg.common.habitica.views.PixelArtView

class PurchaseDialogQuestContent(context: Context) : PurchaseDialogContent(context) {
    private val binding = DialogPurchaseContentQuestBinding.inflate(context.layoutInflater, this)
    override val imageView: PixelArtView
        get() = binding.imageView
    override val titleTextView: TextView
        get() = binding.titleTextView

    override fun setQuestContentItem(questContent: QuestContent) {
        super.setQuestContentItem(questContent)
        binding.rageMeterView.visibility = View.GONE
        if (questContent.isBossQuest) {
            binding.questTypeTextView.setText(R.string.boss_quest)
            binding.questCollectView.visibility = View.GONE
            binding.bossHealthText.text = questContent.boss?.hp.toString()
            if (questContent.boss?.hasRage == true) {
                binding.rageMeterView.visibility = View.VISIBLE
            }
            binding.questDifficultyView.rating = questContent.boss?.str ?: 1f
        } else {
            binding.questTypeTextView.setText(R.string.collection_quest)
            val collectionList = questContent.collect?.map { it.count.toString() + " " + it.text }
            binding.questCollectText.text = TextUtils.join(", ", collectionList ?: listOf<String>())

            binding.bossHealthView.visibility = View.GONE

            binding.questDifficultyView.rating = 1f
        }

        binding.questDetailView.visibility = View.VISIBLE

        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as? LayoutInflater

        if (questContent.drop != null && questContent.drop?.items != null) {
            questContent.drop?.items
                ?.filterNot { it.onlyOwner }
                ?.forEach { addRewardsRow(inflater, it, binding.rewardsList) }

            var hasOwnerRewards = false
            for (item in questContent.drop?.items ?: emptyList<QuestDropItem>()) {
                if (item.onlyOwner) {
                    addRewardsRow(inflater, item, binding.ownerRewardsList)
                    hasOwnerRewards = true
                }
            }
            if (!hasOwnerRewards) {
                binding.ownerRewardsTitle.visibility = View.GONE
                binding.ownerRewardsList.visibility = View.GONE
            }

            if ((questContent.drop?.exp ?: 0) > 0) {
                val view =
                    inflater?.inflate(
                        R.layout.row_quest_reward_imageview,
                        binding.rewardsList,
                        false
                    ) as? ViewGroup
                val imageView = view?.findViewById<ImageView>(R.id.imageView)
                imageView?.scaleType = ImageView.ScaleType.CENTER
                imageView?.setImageBitmap(HabiticaIconsHelper.imageOfExperienceReward())
                val titleTextView = view?.findViewById<TextView>(R.id.titleTextView)
                titleTextView?.text =
                    context.getString(R.string.experience_reward, questContent.drop?.exp)
                binding.rewardsList.addView(view)
            }

            if ((questContent.drop?.gp ?: 0) > 0) {
                val view =
                    inflater?.inflate(
                        R.layout.row_quest_reward_imageview,
                        binding.rewardsList,
                        false
                    ) as? ViewGroup
                val imageView = view?.findViewById<ImageView>(R.id.imageView)
                imageView?.scaleType = ImageView.ScaleType.CENTER
                imageView?.setImageBitmap(HabiticaIconsHelper.imageOfGoldReward())
                val titleTextView = view?.findViewById<TextView>(R.id.titleTextView)
                titleTextView?.text = context.getString(R.string.gold_reward, questContent.drop?.gp)
                binding.rewardsList.addView(view)
            }
        }
    }

    private fun addRewardsRow(
        inflater: LayoutInflater?,
        item: QuestDropItem,
        containerView: ViewGroup?
    ) {
        val view = inflater?.inflate(R.layout.row_quest_reward, containerView, false) as? ViewGroup
        val imageView = view?.findViewById(R.id.imageView) as? PixelArtView
        val titleTextView = view?.findViewById(R.id.titleTextView) as? TextView
        imageView?.loadImage(item.imageName)
        if (item.count > 1) {
            titleTextView?.text =
                context.getString(R.string.quest_reward_count, item.text, item.count)
        } else {
            titleTextView?.text = item.text
        }
        containerView?.addView(view)
    }
}
